src/module.c

```diff
@@ -10388,6 +10393,154 @@ int RM_GetDbIdFromDefragCtx(RedisModuleDefragCtx *ctx) {
     return ctx->dbid;
 }
 
+/* --------------------------------------------------------------------------
+ * ## Modules EventLoop API
+ * --------------------------------------------------------------------------*/
+
+typedef struct EventLoopData {
+    RedisModuleEventLoopCallback rProc;
+    RedisModuleEventLoopCallback wProc;
+    void *user_data;
+} EventLoopData;
+
+static int eventLoopToAeMask(int mask) {
+    int aeMask = 0;
+    if (mask & REDISMODULE_EVENTLOOP_READABLE)
+        aeMask |= AE_READABLE;
+    if (mask & REDISMODULE_EVENTLOOP_WRITABLE)
+        aeMask |= AE_WRITABLE;
+    return aeMask;
+}
+
+static int eventLoopFromAeMask(int ae_mask) {
+    int mask = 0;
+    if (ae_mask & AE_READABLE)
+        mask |= REDISMODULE_EVENTLOOP_READABLE;
+    if (ae_mask & AE_WRITABLE)
+        mask |= REDISMODULE_EVENTLOOP_WRITABLE;
+    return mask;
+}
+
+static void eventLoopCbReadable(struct aeEventLoop *ae, int fd, void *user_data, int ae_mask) {
+    UNUSED(ae);
+    EventLoopData *data = user_data;
+    data->rProc(fd, data->user_data, eventLoopFromAeMask(ae_mask));
+}
+
+static void eventLoopCbWritable(struct aeEventLoop *ae, int fd, void *user_data, int ae_mask) {
+    UNUSED(ae);
+    EventLoopData *data = user_data;
+    data->wProc(fd, data->user_data, eventLoopFromAeMask(ae_mask));
+}
+
+/* Add a pipe / socket event to the event loop
+ *
+ * On success REDISMODULE_OK is returned, otherwise
+ * REDISMODULE_ERR is returned and errno is set to the following values:
+ *
+ * * ERANGE: 'fd' is negative or higher than 'maxclients' Redis config.
+ * * EINVAL: 'callback' is NULL or 'mask' value is invalid.
+ *   errno might take other values in case of an internal error.
+ *
+ *  Example:
+ *  void onReadable(int fd, void *user_data, int mask) {
+ *      char buf[32];
+ *      int bytes = read(fd,buf,sizeof(buf));
+ *      printf("Read %d bytes \n", bytes);
+ *  }
+ *  RM_EventLoopAdd(fd, REDISMODULE_EVENTLOOP_READABLE, onReadable, NULL);
+ */
+int RM_EventLoopAdd(int fd, int mask, RedisModuleEventLoopCallback callback, void *user_data) {
+    if (fd < 0 || fd >= aeGetSetSize(server.el)) {
+        errno = ERANGE;
+        return REDISMODULE_ERR;
+    }
+
+    if (!callback || mask & ~(REDISMODULE_EVENTLOOP_READABLE |
+                              REDISMODULE_EVENTLOOP_WRITABLE)) {
+        errno = EINVAL;
+        return REDISMODULE_ERR;
+    }
+
+    /* We are going to register stub callbacks to 'ae' for two reasons :
+     * - "ae" callback signature is different from RedisModuleEventLoopCallback,
```

Comment:
Must have blank line before bullet list. Otherwise, the text just continues as the same line as “... reasons: - "ae" callback ...”. Also no space before “:” in English.

```suggestion
    /* We are going to register stub callbacks to 'ae' for two reasons:
     *
     * - "ae" callback signature is different from RedisModuleEventLoopCallback,
```